dmoj/commands/validate.py

```diff
@@ -0,0 +1,179 @@
+
+        return total_fails
+
+    def validate_problem(self, problem_id: str) -> bool:
+        print_ansi(f'Validating problem #ansi[{problem_id}](cyan|bold)...')
+
+        problem_root = get_problem_root(problem_id)
+        if problem_root is None:
+            print_ansi(f'\t#ansi[Skipped](magenta|bold) - Problem [{problem_id}](cyan|bold) not found')
+            return True
+
+        assert problem_root is not None
+        config = ProblemConfig(ProblemDataManager(problem_root))
+        if not config or 'validator' not in config:
+            print_ansi('\t#ansi[Skipped](magenta|bold) - No validator found')
+            return True
+
+        validator_config = config['validator']
+        language = validator_config['language']
+        if language not in all_executors:
+            print_ansi('\t\t#ansi[Skipped](magenta|bold) - Language not supported')
+            return True
+
+        time_limit = validator_config.get('time', env.validator_time_limit)
+        memory_limit = validator_config.get('memory', env.validator_memory_limit)
+        compiler_time_limit = validator_config.get('compiler_time_limit', env.validator_compiler_time_limit)
+        read_feedback_from = validator_config.get('read_feedback_from', 'stderr')
+        if read_feedback_from not in ('stdout', 'stderr'):
+            print_ansi('\t\t#ansi[Failed](red|bold) - Feedback option should be (stdout, stderr)')
+            return False
+
+        if isinstance(validator_config.source, str):
+            filenames = [validator_config.source]
+        elif isinstance(validator_config.source.unwrap(), list):
+            filenames = list(validator_config.source.unwrap())
+        else:
+            print_ansi('\t#ansi[Failed](red|bold) - No validator found')
+            return False
+
+        filenames = [os.path.abspath(os.path.join(problem_root, name)) for name in filenames]
+        try:
+            executor = compile_with_auxiliary_files(filenames, lang=language, compiler_time_limit=compiler_time_limit)
+        except CompileError as compilation_error:
+            error_msg = compilation_error.message
+            print_ansi('#ansi[Failed compiling validator!](red|bold)')
+            print(error_msg.rstrip())
+            return False
+
+        problem = Problem(problem_id, time_limit, memory_limit, {})
+        # We need to override the grader's own compilation routine,
+        # because they don't allow for easy multi-file compilation
+        # (e.g. testlib.h) and we also don't know if the grader has already
+        # overwritten the compilation routine (e.g. signature grading, custom grader).
+        # The downside is validating signature graded problems requires
+        # manually setting up the entry point by hand.
+        grader_class = type('Grader', (problem.grader_class,), {'_generate_binary': lambda *_: None})
+        grader = grader_class(self.judge, problem, language, b'')
```

Comment:
Can I get a walkthrough of why the code works like this? The way I imagine validators working is we get the input data, and run the executor directly on the input. I don't feel like using a Grader is important here... what logic are we making use of? Just the case fetching? Can we extract that some other way?

As is, I'm worried that this is a hack that will bite us later.